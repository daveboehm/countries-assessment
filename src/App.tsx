import React, { useState, useEffect } from 'react';
import { fetchAllCountriesData } from './service/countries';
import { CountriesList } from './components/countries-list';
import { FormattedCountriesType } from './types';
import './styles/styles.scss';

function App() {
  const [isLoading, setIsLoading] = useState(true);
  const [countriesList, setCountriesList] = useState<FormattedCountriesType[]>();

  useEffect(() => {
    const getData = async () => {
      try {
        const response: FormattedCountriesType[] = await fetchAllCountriesData();
        setCountriesList(response);
        setIsLoading(false);
      } catch (err) {
        console.error('Error fetching country data', err);
        setIsLoading(false);
      }
    };
    getData();
  }, []);

  return <div>{isLoading ? <p>Imagine this is a beautiful spinner</p> : <CountriesList data={countriesList} />}</div>;
}

export default App;
