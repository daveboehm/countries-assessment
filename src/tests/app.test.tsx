import React from 'react';
import { waitFor, render, RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from '../App';

userEvent.setup();

describe('<App />', () => {
  let view: RenderResult;

  beforeEach(async () => {
    view = render(<App />);

    await waitFor(() => {
      expect(screen.queryAllByRole('row')).toHaveLength(9); // Country headers, 3 countries in list, 5 in details panel
    });
  });

  it('should match snapshot', async () => {
    expect(view.container).toMatchSnapshot();
  });

  it('should filter rows', async () => {
    const filterInput = await screen.findByTestId('filter-input');
    await userEvent.type(filterInput, 'Can');

    const canadaRow = await screen.findByText(/Canada/i);

    expect(canadaRow).toBeVisible();
    await expect(screen.findByText(/Belgium/i)).rejects.toThrow();
  });

  it('should sort by Country name by default', async () => {
    const rows = await screen.findAllByRole('row');

    // row[0] is the table header
    expect(rows[1]).toHaveTextContent(/Belgium/i);
    expect(rows[2]).toHaveTextContent(/Canada/i);
    expect(rows[3]).toHaveTextContent(/Papua New Guinea/i);
  });

  it('should toggle A-Z, Z-A sorting on column header click', async () => {
    const countryCodeButton = screen.getByRole('button', { name: 'Code' });

    // A-Z first click
    await userEvent.click(countryCodeButton);
    await waitFor(() => {
      const rows = screen.getAllByRole('row');
      expect(rows[1]).toHaveTextContent(/BEL/i);
      expect(rows[2]).toHaveTextContent(/CAN/i);
      expect(rows[3]).toHaveTextContent(/PNG/i);
    });

    // Z-A second click
    await userEvent.click(countryCodeButton);
    await waitFor(() => {
      const rows = screen.getAllByRole('row');
      expect(rows[1]).toHaveTextContent(/PNG/i);
      expect(rows[2]).toHaveTextContent(/CAN/i);
      expect(rows[3]).toHaveTextContent(/BEL/i);
    });
  });

  it('should sort population column', async () => {
    const countryPopulationButton = screen.getByRole('button', { name: 'Population' });
    await userEvent.click(countryPopulationButton);

    await waitFor(() => {
      const rows = screen.getAllByRole('row');
      expect(rows[1]).toHaveTextContent(/400/i);
      expect(rows[2]).toHaveTextContent(/500/i);
      expect(rows[3]).toHaveTextContent(/1,000/i);
    });
  });

  it("should display selected country's details", async () => {
    const canadaButton = await screen.findByRole('button', { name: 'Canada' });
    await userEvent.click(canadaButton);

    const detailsPanel = await screen.findByTestId('details-panel');
    expect(detailsPanel).toHaveTextContent(/Ottawa/i);
    expect(detailsPanel).toHaveTextContent(/CAD/i);
    expect(detailsPanel).toHaveTextContent(/883,395/i);
    expect(view.container).toMatchSnapshot();
  });
});
