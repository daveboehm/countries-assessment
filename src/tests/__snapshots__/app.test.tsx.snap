// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should display selected country's details 1`] = `
<div>
  <div>
    <div
      class="countries-list-wrapper"
    >
      <input
        class="filter"
        data-testid="filter-input"
        placeholder="Filter by country name or code"
        type="text"
      />
      <div
        class="data-display-wrapper"
      >
        <div
          class="scroller"
        >
          <table
            class="table"
          >
            <thead>
              <tr
                class="headers row"
              >
                <th
                  class="col"
                >
                  <div
                    class="sortable-header"
                  >
                    <button
                      class="active-sort"
                      type="button"
                    >
                      Name
                       
                      <span
                        class="arrow down"
                      />
                    </button>
                  </div>
                </th>
                <th
                  class="col"
                >
                  <div
                    class="sortable-header"
                  >
                    <button
                      class=""
                      type="button"
                    >
                      Code
                       
                      <span
                        class="arrow"
                      />
                    </button>
                  </div>
                </th>
                <th
                  class="col"
                >
                  <div
                    class="sortable-header"
                  >
                    <button
                      class=""
                      type="button"
                    >
                      Population
                       
                      <span
                        class="arrow"
                      />
                    </button>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="row "
              >
                <td
                  class="col countryName"
                >
                  <button
                    class="show-details-button"
                    type="button"
                  >
                    Belgium
                  </button>
                </td>
                <td
                  class="col countryCode"
                >
                  BEL
                </td>
                <td
                  class="col population"
                >
                  1,000
                </td>
              </tr>
              <tr
                class="row selected"
              >
                <td
                  class="col countryName"
                >
                  <button
                    class="show-details-button"
                    type="button"
                  >
                    Canada
                  </button>
                </td>
                <td
                  class="col countryCode"
                >
                  CAN
                </td>
                <td
                  class="col population"
                >
                  500
                </td>
              </tr>
              <tr
                class="row "
              >
                <td
                  class="col countryName"
                >
                  <button
                    class="show-details-button"
                    type="button"
                  >
                    Papua New Guinea
                  </button>
                </td>
                <td
                  class="col countryCode"
                >
                  PNG
                </td>
                <td
                  class="col population"
                >
                  400
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="country-details-panel"
          data-testid="details-panel"
        >
          <table
            class="country-details"
          >
            <tbody>
              <tr>
                <td>
                  <b>
                    Country Name:
                  </b>
                </td>
                <td>
                  Canada
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Country Code:
                  </b>
                </td>
                <td>
                  CAN
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Capital City:
                  </b>
                </td>
                <td>
                  Ottawa
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Capital Population:
                  </b>
                </td>
                <td>
                  883,395
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Currency:
                  </b>
                </td>
                <td>
                  CAD
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<App /> should match snapshot 1`] = `
<div>
  <div>
    <div
      class="countries-list-wrapper"
    >
      <input
        class="filter"
        data-testid="filter-input"
        placeholder="Filter by country name or code"
        type="text"
      />
      <div
        class="data-display-wrapper"
      >
        <div
          class="scroller"
        >
          <table
            class="table"
          >
            <thead>
              <tr
                class="headers row"
              >
                <th
                  class="col"
                >
                  <div
                    class="sortable-header"
                  >
                    <button
                      class="active-sort"
                      type="button"
                    >
                      Name
                       
                      <span
                        class="arrow down"
                      />
                    </button>
                  </div>
                </th>
                <th
                  class="col"
                >
                  <div
                    class="sortable-header"
                  >
                    <button
                      class=""
                      type="button"
                    >
                      Code
                       
                      <span
                        class="arrow"
                      />
                    </button>
                  </div>
                </th>
                <th
                  class="col"
                >
                  <div
                    class="sortable-header"
                  >
                    <button
                      class=""
                      type="button"
                    >
                      Population
                       
                      <span
                        class="arrow"
                      />
                    </button>
                  </div>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr
                class="row "
              >
                <td
                  class="col countryName"
                >
                  <button
                    class="show-details-button"
                    type="button"
                  >
                    Belgium
                  </button>
                </td>
                <td
                  class="col countryCode"
                >
                  BEL
                </td>
                <td
                  class="col population"
                >
                  1,000
                </td>
              </tr>
              <tr
                class="row "
              >
                <td
                  class="col countryName"
                >
                  <button
                    class="show-details-button"
                    type="button"
                  >
                    Canada
                  </button>
                </td>
                <td
                  class="col countryCode"
                >
                  CAN
                </td>
                <td
                  class="col population"
                >
                  500
                </td>
              </tr>
              <tr
                class="row "
              >
                <td
                  class="col countryName"
                >
                  <button
                    class="show-details-button"
                    type="button"
                  >
                    Papua New Guinea
                  </button>
                </td>
                <td
                  class="col countryCode"
                >
                  PNG
                </td>
                <td
                  class="col population"
                >
                  400
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="country-details-panel"
          data-testid="details-panel"
        >
          <table
            class="country-details"
          >
            <tbody>
              <tr>
                <td>
                  <b>
                    Country Name:
                  </b>
                </td>
                <td>
                  -
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Country Code:
                  </b>
                </td>
                <td>
                  -
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Capital City:
                  </b>
                </td>
                <td>
                  -
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Capital Population:
                  </b>
                </td>
                <td>
                  0
                </td>
              </tr>
              <tr>
                <td>
                  <b>
                    Currency:
                  </b>
                </td>
                <td>
                  -
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
